br/pkg/stream/stream_mgr.go

```diff
@@ -191,13 +210,39 @@ func (m *MetadataHelper) decodeCompressedData(data []byte, compressionType backu
 		"failed to decode compressed data: compression type is unimplemented. type id is %d", compressionType)
 }
 
+func (m *MetadataHelper) verifyChecksumAndDecryptIfNeeded(ctx context.Context, data []byte, encryptionInfo *encryptionpb.FileEncryptionInfo) ([]byte, error) {
+	// no need to decrypt
+	if encryptionInfo == nil {
+		return data, nil
+	}
+
+	if m.encryptionManager == nil {
+		return nil, errors.New("need to decrypt data but encryption manager not set")
+	}
+
+	// Verify checksum before decryption
+	if encryptionInfo.Checksum != nil {
+		actualChecksum := sha256.Sum256(data)
+		if !bytes.Equal(actualChecksum[:], encryptionInfo.Checksum) {
+			return nil, errors.Errorf("checksum mismatch before decryption, expected %s, actual %s", encryptionInfo.Checksum, actualChecksum)
```

Comment:
Maybe `hex.EncodeToString` here,

```suggestion
			return nil, errors.Errorf("checksum mismatch before decryption, expected %s, actual %s", hex.EncodeToString(encryptionInfo.Checksum), hex.EncodeToString(actualChecksum))
```